File: go/007/007.go
package main

import "fmt"
import "math"

func sieveOfAtkin(limit uint64) []uint64 {
	var primes []uint64

	sieve := make([]bool, limit+1)

	var i uint64

	for i = 0; i < limit; i++ {
		sieve[i] = false
	}

	squareRootOfLimit := uint64(math.Sqrt(float64(limit)))

	var x uint64

	var y uint64

	var n uint64

	for x = 1; x <= squareRootOfLimit; x++ {
		for y = 1; y <= squareRootOfLimit; y++ {
			n := (4 * x * x) + (y * y)

			if n <= limit && (n%12 == 1 || n%12 == 5) {
				sieve[n] = !sieve[n]
			}

			n = (3 * x * x) + (y * y)

			if n <= limit && n%12 == 7 {
				sieve[n] = !sieve[n]
			}

			n = (3 * x * x) - (y * y)

			if x > y && n <= limit && n%12 == 11 {
				sieve[n] = !sieve[n]
			}
		}
	}

	sieve[2] = true

	sieve[3] = true

	primes = append(primes, 2)

	primes = append(primes, 3)

	n = 5

	for n <= squareRootOfLimit {
		if sieve[n] {
			i := n * n

			for i < limit {
				sieve[i] = false

				i += n * n
			}

			primes = append(primes, n)
		}

		n += 2
	}

	for n < limit {
		if sieve[n] {
			primes = append(primes, n)
		}

		n += 2
	}

	return primes
}

func main() {
	n := 10001

	limit := uint64(((float64(n) * math.Log(float64(n))) + (float64(n) * (math.Log(math.Log(float64(n)))))))

	fmt.Println(sieveOfAtkin(limit)[n-1])
}

File: julia/007/007.jl
function sieve_of_atkin(limit)
  primes = Int64[]

  sieve = falses(limit)

  square_root_of_limit = round(Int64, sqrt(limit))

  for x in 1:square_root_of_limit
    for y in 1:square_root_of_limit
      n = (4 * x * x) + (y * y)

      if n <= limit && (n % 12 == 1 || n % 12 == 5)
        sieve[n] = !sieve[n]
      end

      n = (3 * x * x) + (y * y)

      if n <= limit && n % 12 == 7
        sieve[n] = !sieve[n]
      end

      n = (3 * x * x) - (y * y)

      if x > y && n <= limit && n % 12 == 11
        sieve[n] = !sieve[n]
      end
    end
  end

  sieve[2] = true

  sieve[3] = true

  push!(primes, 2)

  push!(primes, 3)

  n = 5

  while n <= square_root_of_limit
    if sieve[n]
      i = n * n

      while i < limit
        sieve[i] = false

        i += n * n
      end

      push!(primes, n)
    end

    n += 2
  end

  while n < limit
    if sieve[n]
      push!(primes, n)
    end

    n += 2
  end

  primes
end

n = 10001

limit = round(Int64, (n * log(n)) + (n * (log(log(n)))))

println(sieve_of_atkin(limit)[n])
